src/Akka.Streams.Kafka/Stages/Consumers/Abstract/BaseSingleSourceLogic.cs

```diff
@@ -146,25 +147,23 @@ protected virtual void MessageHandling((IActorRef, object) args)
 
         private void Pump()
         {
-            if (IsAvailable(_shape.Outlet))
+            while(IsAvailable(_shape.Outlet) && _buffer.TryDequeue(out var message))
             {
-                if (_buffer.TryDequeue(out var message))
-                {
-                    Push(_shape.Outlet, _messageBuilder.CreateMessage(message));
-                    Pump();
-                }
-                else if (!_requested && TopicPartitions.Any())
-                {
-                    RequestMessages();
-                }
+                Push(_shape.Outlet, _messageBuilder.CreateMessage(message));
+            }
+            
+            if (IsAvailable(_shape.Outlet) && !_requested && TopicPartitions.Any())
+            {
+                RequestMessages();
             }
         }
 
         protected void RequestMessages()
         {
             _requested = true;
             _requestId += 1;
-            Log.Debug($"Requesting messages, requestId: {_requestId}, partitions: {string.Join(", ", TopicPartitions)}");
+            if (Log.IsDebugEnabled)
+                Log.Debug($"Requesting messages, requestId: {_requestId}, partitions: {string.Join(", ", TopicPartitions)}");
```

Comment:
Added some debug logging here to help make it easier to understand what each consumer is doing - won't allocate any new strings or events unless `akka.loglevel = DEBUG`